### ISSUE TITLE
About the implementation of the face detector.

### ISSUE THREAD

**Issue by II-Matto** [liuliu/ccv#94]
Hi, I checked out the implementation of face detector in ccv and felt kind of confused.

The bbf page references two papers, the latter of which is _High-Performance Rotation Invariant Multiview Face Detection_ (**PAPER 2**). So I thought the face detector is implemented as described, i.e. having WFS tree structure, using sparse granular features, and based on vector boosting.

What I found confused is that the code seems to implement a simple cascade structure rather than a tree structure. There is an outer loop indexed by `i` corresponding to layers of the cascade, and an inner loop indexed by `k` corresponding to features selected. I didn't see a clue of tree structure if I didn't miss anything.

And moreover, taking a look at the data structures for the feature, I suppose it is not the sparse granular feature used in **PAPER 2** that is implemented, but the control points feature in the other paper, which enlightened me about the name "Brightness Binary Feature".

I'm not sure if I took anything wrong, or it is true that the code didn't implement the vector boosting tree-structured face detector.

Besides, I tested the shipped detector on MIT-CMU face image set. The highest recall obtained is **77.5%** and the false alarm **6**, which is different from the results presented on the bbf page. There are also problems with the evaluation method in _bbfvldtr.rb_, because a detection box may enclose multiple faces, which should be taken as a false alarm rather than a true positive.

**Comment by liuliu** [comment]
You are right on both. The WFS is not implemented here, but as a suggestion for improvement. I'd argue that sparse granular feature is the same as control points feature, besides its adoption to the specific boosting method used.

I will modify the bbfvldtr.rb. Good catch on this!

This face detector is not particular good (as comparing with state-of-the-art). I will have another face detector implemented at some point. Also, MIT-CMU should be deprecated as test set as well. There are plans for ccv to include more recent result and comparison with more state-of-the-art (ImageNet 2012, VOC 2012, and FDDB).

**Comment by II-Matto** [comment]
As for the control-points feature, the **PAPER 2** cited the corresponding paper with the following remark:

> As a result, these pixel-based features are extremely fast to compute but unfortunately not discriminative and robust enough. In our face detection system, a set of novel features are sparsely represented in the granular space of the input image, ...

The so-called pixel-based features include the control-points feature, so I suppose it could not be the **same** as the **novel** features. At least the sparse granular features are not binary. I agree that neither of the two types of features are constrained by a specific variant of boosting method.

The MIT-CMU face dataset is indeed obsolete. I also tested the face detector on FDDB. The results, say, are not so good. The detection rate is under **20%** with **100** false alarms, while the highest rate is less than **60%**. The evaluation was done with the code provided by FDDB officials. But to be honest, this result is partially due to the evaluation scheme. The output bounding box of ccv face detector is a small square, while the FDDB annotations are large ellipses. Consequently, the intersection-union ratios tend to be small. Besides, FDDB is more challenging than MIT-CMU and LFW dataset. The following figure shows the ROC curve.

![roc](https://cloud.githubusercontent.com/assets/2261229/3295321/08d239e4-f5bd-11e3-9444-10706aa646cd.png)

**Comment by liuliu** [comment]
That's not surprising, the ccv's implementation is not in-plane-rotation invariant. I am leaning towards a SURF-cascade based method to replace existing one: https://sites.google.com/site/leeplus/publications/learningsurfcascadeforfastandaccurateobjectdetection now because it requires less training data and has good result on FDDB.
There are at least 7 to 10 years gap of ccv's implementation to the state-of-the-art. I am actually working on a graph to show the time gap between open-source implementations (of computer vision) and state-of-the-art.

**Comment by II-Matto** [comment]
That seems appealing! Really look forward to the new implementation.

**Comment by II-Matto** [comment]
Hi, the bbf document page mentioned that face regions are "_cropped with random rotation, translation and scale variations_". I was wondering how, exactly, it is done. I come up with two possible ways for rotation.

(1) Crop the face from the image (a tight face region, large enough to enclose the eyes and mouth, with some space around, i.e. few background pixels), and then rotate the face image to get more. Empty pixels are filled with 0.
(2) Rotate the original image about the center of the face rectangle, i.e. the face rectangle is fixed. Then crop from rotated images to get new faces. In this case, there generally should be no empty pixels.

The former way may not be preferable since the zero-filled empty pixels will degrade the accuracy of the face detector. So which is the one you employed or you actually choose a third one? Besides, if I only have cropped faces at hand, what do you think is the best way to get rotated-face images so that the robustness of trained detector is truly boosted rather than degraded?

BTW, kind of confused about the translation and scale variations. Shouldn't the face be approximately in the center of the cropped image and they all be resized into 24x24?

**Comment by liuliu** [comment]
I cropped after rotation.

translation and scale is normally used to generate these examples to combat overfitting (actually pretty common with these excessive feature pool). it is first applying rotation, translation and scaling, and then cropped to 24x24.
